### PR TITLE
API improvements.

### LINKED ISSUE #2: XML documentation

### PR DESCRIPTION
**Pull request by vcsjones**
@onovotny

Take a look at the new program sample. What do you think of this as a start?

This addresses:

- A somewhat friendlier API for going through the signature tree
- Fixes the crash because I was casting incorrectly when extracting the timestamp
- Correctly shows the timestamp for RFC3161 timestamps

```csharp
var extractor = new FileInspector(path);

//Returns an enum of the validation result
var validationResult = extractor.Validate(); 

// Gets all of the signatures "flattened", no recursion necessary. 
var signatures = extractor.GetSignatures();

foreach(var signature in signatures)
{
    //Property to get the publisher info, or null
    //Signtool always puts in a publisherinfo, the fields are just empty strings.
    var publisherInfo = signature.PublisherInformation;

    //Property to get the timestamp signature
    var timestamps = signature.TimestampSignatures;

     // Gets the DateTimeOffset of the timestamp
    var timestampedAt = timestamp.TimestampDateTime;
}
```

Fixes #2

### CODE REVIEW

**Comment by clairernovotny** [comment]
That's better, but what do you think of something closer to https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/view-authenticode-sigs/PackageViewModel/Types/AuthenticodeSignature.cs

It has a signature object that directly exposes the publisher info, timestamp info, etc?

Null is valid if there isn't one.

**Comment by vcsjones** [comment]
The thought was that since all of signature types are the same interface, and in the case of RF3161 timestamps they're actually the same class as a normal signature, It doesn't make sense for timestamps to have publisher info on them. That was the thinking at least.

I can encapsulate the known signatures to be more "concrete" as to what they are and expose things on the types rather than extensions.

A small issue with my current design I intend to correct is that a signature can have multiple time stamps (which I've seen in the wild.. rarely) so Timestamps should be a collection.

**Comment by clairernovotny** [comment]
Yeah, was thinking about keeping it as simple for people that might not know what kinds of things are in a sig. That an object would represent all of that.

**Comment by vcsjones** [comment]
@onovotny updated

**Comment by clairernovotny** [comment]
Looks better, but still not clear to me why Publisher and Timestamp should be methods? Seems like they could be calculated quickly either on-demand on first access or in the ctor? That also makes it possible to bind in WPF.

**Comment by vcsjones** [comment]
>That also makes it possible to bind in WPF.

Ah. Wasn't aware you were going to use these as models.

**Comment by clairernovotny** [comment]
No reason to write a wrapper model if it's not necessary :)

**Comment by vcsjones** [comment]
okie doke, properties it is.

**Comment by clairernovotny** [comment]
:shipit:

File: src/AuthenticodeExaminer/PublisherInformation.cs [hunk 1]
```diff
@@ -26,6 +26,12 @@ public class PublisherInformation
         /// </summary>
         public string FileLink { get; }
 
+        /// <summary>
+        /// Constructs an empty instance of <see cref="PublisherInformation"/>.
+        /// </summary>
+        public PublisherInformation()
+        {
```
Comment (clairernovotny) [inline src/AuthenticodeExaminer/PublisherInformation.cs:42]: Should there be an `IsEmpty` property since it'll never be null?

Comment (vcsjones) [reply src/AuthenticodeExaminer/PublisherInformation.cs:42]: @onovotny Would `IsEmpty` only be true if the default constructor ends up getting used, or something more like:

```csharp
public bool IsEmpty =>
    string.IsNullOrWhiteSpace(Description) &&
    string.IsNullOrWhiteSpace(UrlLink) &&
    string.IsNullOrWhiteSpace(FileLink);
```

In practice, signtool always creates a PublisherInformation (SPC_SP_OPUS_INFO) and embeds it, and the Description and Url contain a single whitespace character, even if you don't pass the `-d` or `-du` parameter.

Should we try to normalize that somehow?

Comment (clairernovotny) [reply src/AuthenticodeExaminer/PublisherInformation.cs:42]: I think it'd make sense to normalize it the way you suggest.

**Comment by vcsjones** [comment]
@onovotny give https://www.nuget.org/packages/AuthenticodeExaminer/0.2.0 a shot when you got a moment.

**Pull request merged**